File: various/GetOpt.hs.pl
use strict;

# scanning mode: one of "at_beginning", "anywhere", "all_options"
# kind: one of "noarg", "arg", "optional_arg"

my ($SHORT, $LONG, $KIND, $DESCR) = (0 .. 3);

sub usageInfo {
    my ($opt_descr) = @_;

    my @l = map { [ 
		   fmt_opt(0, $_->[$KIND], split('', $_->[$SHORT])),
		   fmt_opt(1, $_->[$KIND], split(' ', $_->[$LONG])),
		   $_->[$DESCR]
		  ] } @$opt_descr;

    my @max_sizes = map { maximum(map length, @$_) } unzip(@l);

    map {
	my $i;
	join('', map { "  $_" . " " x ($max_sizes[$i++] - length) } @$_);
    } @l;
}

sub fmt_opt {
    my ($is_long, $kind, @opts) = @_;

    my ($dash, $space, $equal) = $is_long ? ("--", "", "=") : ("-", " ", "");

    my @l = do {
	if ($kind =~ /^arg=(.*)/) {
	    map { "$dash$_$space$equal$1" } @opts;
	} elsif ($kind =~ /optional_arg=(.*)/) {
	    map { "$dash$_\[$equal$1\]" } @opts;
	} else {
	    map { "$dash$_" } @opts;
	}
    };
    join(", ", @l);
}

sub get_opt {
    my ($opt_descr, $scanning_mode, @args) = @_;
    
    my (@args_in_between, @pbs, @options);

    my $check_matches = sub {
	my ($opt, @matches) = @_;
	if (@matches == 1) {
	    $matches[0];
	} else {
	    push @pbs, @matches ? 
	      ("option `$opt' is ambiguous; could be one of:", usageInfo(\@matches)) :
	      "unrecognized option `$opt'";
	    '';
	}
    };

    my $handle_long = sub {
	my ($opt, $arg) = @_;

	my $match = $check_matches->($opt, grep { $_->[$LONG] =~ /\b\Q$opt/ } @$opt_descr) or return;
	my $kind = $match->[$KIND];
	
	if ($kind eq 'noarg' && $arg) {
	    push @pbs, "option `$opt' doesn't allow an argument";
	} elsif ($kind =~ /^arg=(.*)/ && !$arg) {
	    push @pbs, "option `$opt' requires an argument $1";
	} else {
	    push @options, [ $match->[$LONG], $arg ];
	}
    };
    my $handle_short = sub {
	my ($opts) = @_;

	while ($opts) {
	    (my $opt, $opts) = $opts =~ /(.)(.*)/;

	    my $match = $check_matches->($opt, grep { $_->[$SHORT] =~ /\Q$opt/ } @$opt_descr) or next;
	    my $kind = $match->[$KIND];

	    my $arg;
	    if ($kind eq 'noarg') {
	    } elsif ($opts) {
		($arg, $opts) = ($opts, '');
	    } elsif ($kind =~ /^arg=(.*)/) {
		$arg = shift @args
		  or push(@pbs, "option `$opt' requires an argument $1"), return;
	    }
	    push @options, [ $match->[$LONG], $arg ];
	}
    };

    while (@args) {
	local $_ = shift @args;
	if    (/^--$/)         { last } # if $scanning_mode ne 'all_options'
	elsif (/^--(.*)=(.*)/) { $handle_long->($1, $2) }
	elsif (/^--(.*)/)      { $handle_long->($1) }
	elsif (/^-(.+)/)       { $handle_short->($1) }
	else {
	    push @args_in_between, $_;
	    last if $scanning_mode eq 'at_beginning';
	}
    }
    @pbs and die join("\n", @pbs, '');

    if ($scanning_mode eq 'all_options') {
	push @options, [ 'Arg', $_ ] foreach @args_in_between;
	@args_in_between = '';
    }
    \@options, [ @args_in_between, @args ];
}

############################################-
# and here a small and hopefully enlightening example:

my @opt_descr = (
    [ 'v',  'verbose',         'noarg',             "verbosely list files" ],
    [ 'V?', 'version release', 'noarg',             "show version info" ],
    [ 'o',  'output',          'optional_arg=FILE', "use FILE for dump" ],
    [ 'n',  'name',            'arg=USER',          "only dump USER's files" ],
);

eval {
    my ($options, $args) = get_opt(\@opt_descr, at_beginning => qw(-?));
    use Data::Dumper;
    print Dumper($options, $args);
};
if ($@) {
    die join("\n", $@ . "Usage: foobar [OPTION...] files...", usageInfo(\@opt_descr), '');
}


sub maximum { my $n = shift; $_ > $n and $n = $_ foreach @_; $n }
sub unzip {
    my @r;
    foreach (@_) {
	my $i;
	push @{$r[$i++]}, $_ foreach @$_;
    }
    @r;
}

File: various/GetOpt.hs.rb
def mapn(*ll)
  ll.empty? and return []
  r = []
  ll[0].each_index{|i| r << yield(ll.map{|l| l[i]})}
  r
end
def unzip(l); mapn(*l){|l| l}; end

# scanning mode: one of "at_beginning", "anywhere", "all_options"
# kind: one of "noarg", "arg", "optional_arg"

Options = Struct.new("Options", :short, :long, :kind, :descr)

def usageInfo(opt_descr)
  ll = opt_descr.map{|opts|
    [
      fmt_opt(false, opts.kind, opts.short),
      fmt_opt(true,  opts.kind, opts.long.split),
      opts.descr,
    ]
  }

  max_sizes = unzip(ll).map{|l| l.map{|e| e.length}.max}

  ll.map{|l|
    mapn(l, max_sizes){|e, max_size| 
      "  " + e + " " * (max_size - e.length)
    }.join
  }.join("\n")
end

def fmt_opt(is_long, kind, opts)
  (dash, space, equal) = is_long ? ["--", "", "="] : ["-", " ", ""]

  l = 
    if kind =~ /^arg=(.*)/
      opts.map{|e| dash + e + space + equal + $1}
    elsif kind =~ /optional_arg=(.*)/
      opts.map{|e| dash + e + "[" + equal + $1 + "]" }
    else
      opts.map{|e| dash + e}
    end
  l.join(", ")
end

def get_opt(opt_descr, scanning_mode, *args)
    
  args_in_between, pbs, options = [], [], []

  check_matches = proc{|opt, matches|
    if matches.size == 1
      matches[0]
    else
      pbs << 
	if matches.empty?
	  "unrecognized option `#{opt}'";
	else
	  "option `#{opt}' is ambiguous; could be one of:\n" + usageInfo(matches)
	end
      false
    end
  }

  handle_long = proc{|opt, arg|
    match = check_matches.call(opt, opt_descr.find_all{|e| e.long =~ /\b#{Regexp.quote(opt)}/ }) or break
    kind = match.kind
	
    if kind == 'noarg' && arg
      pbs << "option `#{opt}' doesn't allow an argument"
    elsif kind =~ /^arg=(.*)/ && !arg
      pbs << "option `#{opt}' requires an argument 1"
    else
      options << [ match.long, arg ]
    end
  }
  handle_short = proc{|opts|
    while opts != ""
      opt = opts.slice!(0..0)

      match = check_matches.call(opt, opt_descr.find_all{|e| e.short.include?(opt) }) or next
      kind = match.kind

      if kind == 'noarg'
	arg = ''
      elsif opts
	arg, opts = opts, ''
      elsif kind =~ /^arg=(.*)/
	arg = args.shift or pbs << "option `#{opt}' requires an argument #{$1}" ; break
      end
      options << [ match.long, arg ]
    end
  }

  while args
    $_ = args.shift

    if    ~ /^--$/;         break # if scanning_mode ne 'all_options'
    elsif ~ /^--(.*)=(.*)/; handle_long.call($1, $2)
    elsif ~ /^--(.*)/;      handle_long.call($1, nil)
    elsif ~ /^-(.+)/ ;      handle_short.call($1)
    else 
      args_in_between << $_
      break if scanning_mode == 'at_beginning'
    end
  end
  pbs.empty? or raise pbs.join("\n")

  if scanning_mode == 'all_options'
    options << args_in_between.map{|e| [ 'Arg', e ]}
    args_in_between = []
  end
  return options, args_in_between + args
end

############################################-
# and here a small and hopefully enlightening example:

opt_descr = [
    Options.new('v',  'verbose',         'noarg',             "verbosely list files"),
    Options.new('V?', 'version release', 'noarg',             "show version info"),
    Options.new('o',  'output',          'optional_arg=FILE', "use FILE for dump"),
    Options.new('n',  'name',            'arg=USER',          "only dump USER's files"),
]

begin
  p get_opt(opt_descr, "at_beginning", '-V', 'ee')
rescue
  print "#{$!}\n" + "Usage: foobar [OPTION...] files...\n" + usageInfo(opt_descr)
end
